zio-schema/shared/src/main/scala/zio/schema/MetaSchema.scala

```diff
@@ -0,0 +1,128 @@
+package zio.schema
+
+import java.time.temporal.{ ChronoUnit, TemporalUnit }
+
+import scala.collection.immutable.ListMap
+
+import zio.Chunk
+
+sealed trait MetaSchema { self =>
+
+  def toSchema: Schema[_] = self match {
+    case MetaSchema.Duration(units)         => Schema.Primitive(StandardType.Duration(units))
+    case MetaSchema.Value(tpe)              => Schema.Primitive(tpe)
+    case MetaSchema.Fail(message)           => Schema.Fail(message)
+    case MetaSchema.Optional(meta)          => meta.toSchema.optional
+    case MetaSchema.EitherMeta(left, right) => left.toSchema <+> right.toSchema
+    case MetaSchema.TupleMeta(left, right)  => left.toSchema <*> right.toSchema
+    case MetaSchema.Record(structure) =>
+      Schema.GenericRecord(structure.map(metaField => Schema.Field(metaField.label, metaField.fieldType.toSchema)))
+    case MetaSchema.Enum(cases) =>
+      Schema.Enumeration(ListMap.empty ++ cases.map(metaCase => metaCase.id -> metaCase.subType.toSchema))
+    case MetaSchema.Sequence(tpe) => Schema.chunk(tpe.toSchema)
+  }
+}
+
+object MetaSchema {
+
+  def fromSchema[A](schema: Schema[A]): MetaSchema = schema match {
+    case Schema.Primitive(StandardType.Duration(units)) => Duration(units)
+    case Schema.Primitive(tpe)                          => Value(tpe)
+    case Schema.Optional(schema)                        => Optional(fromSchema(schema))
+    case Schema.EitherSchema(left, right)               => EitherMeta(fromSchema(left), fromSchema(right))
+    case Schema.Tuple(left, right)                      => TupleMeta(fromSchema(left), fromSchema(right))
+    case Schema.Sequence(schema, _, _)                  => Sequence(fromSchema(schema))
+    case Schema.Fail(message)                           => Fail(message)
+    case Schema.Transform(schema, _, _)                 => fromSchema(schema)
+    case lzy @ Schema.Lazy(_)                           => fromSchema(lzy.schema)
```

Comment:
We should test this on a schema like JSON which is recursive. I think our "schema language" may have to embed references in order to handle recursion (or some sort of "fixed point" operator).

Reply:
Yeah, that is going to be a problem. I can work on this today.

Reply:
The [JSON Schema](https://json-schema.org/specification.html) doc uses "pointers" for recursion.

To track recursion, you can use a map based on object identity. This will allow you to detect if you processed the same field before. It might also be possible to define equals / hashCode on the ordinary `Schema` classes in such a way that it terminates even for recursive structures.